### ISSUE TITLE
brew search stuck

### ISSUE THREAD

**Issue by cscetbon** [Homebrew/homebrew-core#4793]
- [X] Ran `brew update` and retried your prior step?
- [X] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?

### Bug reports:
**brew search** is stuck
```
$ brew tap
homebrew/completions
homebrew/core
homebrew/dupes
homebrew/gui
homebrew/services
homebrew/versions
homebrew/x11
$ brew search curl                                                                                                                                  
curl                               curlish                            curlpp                             flickcurl                          homebrew/x11/curlftpfs
## It is stuck ## here
$ ps -ef|grep brew # in parallel
/usr/bin/curl --progress-bar --remote-time --location --user-agent Homebrew/0.9.9 (Macintosh; Intel Mac OS X 10.11.6) curl/7.43.0 https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD?recursive=1 --header application/vnd.github.v3+json --write-out ^J%{http_code} /usr/bin/curl --progress-bar --remote-time --location --user-agent Homebrew/0.9.9 (Macintosh; Intel Mac OS X 10.11.6) curl/7.43.0 --header Authorization: token xxxxxxxxxx --dump-header /tmp/github_api_headersyyyyyyyy
```
I'm wondering why it tries to access https://api.github.com/repos/Caskroom/homebrew-cask after I have removed **caskroom/cask** using 
```
$ brew untap caskroom/cask
Untapping caskroom/cask... (3,644 files, 68.1M)
Untapped 0 formulae
```
I wanted to reinstall it to try to fix the issue ..

Any idea ?

**Comment by zmwangx** [comment]
`brew search` searches not only locally, but also online, in official taps under the Homebrew organization plus `caskroom/cask` and `caskroom/versions`. Here's how the code works:

- https://git.io/viaSg
- https://git.io/viaS2
- https://git.io/viaSa
- https://git.io/viaSr

**Comment by zmwangx** [comment]
By the way, if the online operations are slow, you can always ^C to kill it. There's no harm.

**Comment by cscetbon** [comment]
Okay but why is it so slow ? I wasn't able to get an answer. Why does it search through caskroom when this tap has been removed ?

**Comment by zmwangx** [comment]
> Why does it search through caskroom when this tap has been removed ?

Have you read what I wrote above? It searches *online* "in official taps under the Homebrew organization plus `caskroom/cask` and `caskroom/versions`". The fact that you removed the tap makes it even slower: it could have searched offline for the particular tap, but now it has to query the API.

> Okay but why is it so slow?

Depending on how many of the official taps you have locally, the search command could make up to seventeen API requests to GitHub. Apply this patch to print them:

```diff
modified   Library/Homebrew/cmd/search.rb
@@ -130,6 +130,7 @@ module Homebrew
       user, repo = key.split("/", 2)
       tree = {}

+      puts "https://api.github.com/repos/#{user}/homebrew-#{repo}/git/trees/HEAD?recursive=1"
       GitHub.open "https://api.github.com/repos/#{user}/homebrew-#{repo}/git/trees/HEAD?recursive=1" do |json|
         json["tree"].each do |object|
           next unless object["type"] == "blob"
```

If your connection to GitHub is slow, you're screwed.

**Comment by zmwangx** [comment]
Feel free to ask if it's still unclear.

**Comment by cscetbon** [comment]
@zmwangx Okay I get it. Sorry I missed the part about the official taps remote processing. I suppose it has been added by last home-brew versions. My Github connection is correct, as I use github everyday without any issue. However I really think that you have an issue with the way you do the search. It takes more than 5 minutes (I've always stopped it manually) to get an answer which is problem. So I don't understand why you close it. How much time does it take you to make a search query ?

**Comment by zmwangx** [comment]
> How much time does it take you to make a search query ?

A couple seconds, 3 when I just tested. It could have been 50% or so longer because I have several official taps offline.

> So I don't understand why you close it.

I answered the question, that's it. I also said "feel free to ask if it's still unclear."

> However I really think that you have an issue with the way you do the search. It takes more than 5 minutes (I've always stopped it manually) to get an answer which is problem.

Apply the aforementioned patch and see what URLs you're connecting to and how fast they are. Do note that you'll run out of quota fast as an anonymous user to the API, so you'll need to set `HOMEBREW_GITHUB_API_TOKEN` for continued testing. See `man brew`.

**Comment by cscetbon** [comment]
I already have a specific Github token; I've regenerated it (just in case) and tried again without any success. Here is the output of the added **puts** in **Library/Homebrew/cmd/search.rb**
```
https://api.github.com/repos/Homebrew/homebrew-devel-only/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-apache/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-nginx/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-science/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-tex/git/trees/HEAD?recursive=1
https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-emacs/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-php/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-games/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-python/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-fuse/git/trees/HEAD?recursive=1
https://api.github.com/repos/Caskroom/homebrew-versions/git/trees/HEAD?recursive=1
```

Should I test all of them manually ? If yes, how ?

**Comment by zmwangx** [comment]
If you just curl them, are they slow to get?

**Comment by zmwangx** [comment]
Note: some of these could be pretty big, the biggest one probably being Caskroom/homebrew-cask:

```
$ wget --spider https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD\?recursive\=1

Spider mode enabled. Check if remote file exists.
--2016-09-14 08:53:56--  https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD?recursive=1
Resolving api.github.com (api.github.com)... 192.30.253.116
Connecting to api.github.com (api.github.com)|192.30.253.116|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 991350 (968K) [application/json]
Remote file exists.

```

That's almost a solid 1MB.

**Comment by cscetbon** [comment]
1 MB is almost nothing for my connection... It takes me a second to download each of them . There is something else that causes this slowness. Any other idea ?

**Comment by zmwangx** [comment]
1 MB may be nothing for your connection, but the process you caught in the process table was indeed the 1 MB download:

```
/usr/bin/curl --progress-bar --remote-time --location --user-agent Homebrew/0.9.9 (Macintosh; Intel Mac OS X 10.11.6) curl/7.43.0 https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD?recursive=1 --header application/vnd.github.v3+json --write-out ^J%{http_code} /usr/bin/curl --progress-bar --remote-time --location --user-agent Homebrew/0.9.9 (Macintosh; Intel Mac OS X 10.11.6) curl/7.43.0 --header Authorization: token xxxxxxxxxx --dump-header /tmp/github_api_headersyyyyyyyy
```

**Comment by cscetbon** [comment]
Not knowing the internal code, I really can't answer ...

**Comment by zmwangx** [comment]
Also try this patch

```diff
diff --git a/Library/Homebrew/cmd/search.rb b/Library/Homebrew/cmd/search.rb
index 50d9f5f..871d24e 100644
--- a/Library/Homebrew/cmd/search.rb
+++ b/Library/Homebrew/cmd/search.rb
@@ -130,6 +130,7 @@ module Homebrew
       user, repo = key.split("/", 2)
       tree = {}
 
+      start_time = Time.now
       GitHub.open "https://api.github.com/repos/#{user}/homebrew-#{repo}/git/trees/HEAD?recursive=1" do |json|
         json["tree"].each do |object|
           next unless object["type"] == "blob"
@@ -142,6 +143,8 @@ module Homebrew
           end
         end
       end
+      end_time = Time.now
+      puts "#{user}/#{repo}: #{end_time - start_time}"
 
       paths = tree["Formula"] || tree["HomebrewFormula"] || tree["."] || []
       paths += tree["Casks"] || []
```

**Comment by zmwangx** [comment]
Sample run of patched version:

```zsh
$ time (brew search wmf)
libwmf
Homebrew/gui: 0.253559
Homebrew/apache: 0.361786
Homebrew/tex: 0.359082
Homebrew/completions: 0.471945
Homebrew/games: 0.602125
Homebrew/x11: 0.634227
Homebrew/nginx: 0.646969
Homebrew/fuse: 0.6625
Homebrew/emacs: 0.899245
Caskroom/versions: 0.979446
Homebrew/php: 1.510015
Caskroom/cask: 6.03597
( brew search wmf; )  2.79s user 0.43s system 49% cpu 6.536 total
```

**Comment by cscetbon** [comment]
```
$ time (brew search wmf)
libwmf
https://api.github.com/repos/Homebrew/homebrew-devel-only/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-apache/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-fuse/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-php/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-science/git/trees/HEAD?recursive=1
https://api.github.com/repos/Caskroom/homebrew-cask/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-emacs/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-games/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-tex/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-python/git/trees/HEAD?recursive=1https://api.github.com/repos/Caskroom/homebrew-versions/git/trees/HEAD?recursive=1
https://api.github.com/repos/Homebrew/homebrew-nginx/git/trees/HEAD?recursive=1

Homebrew/python: 1.489939
Homebrew/tex: 1.50934
Homebrew/nginx: 1.572022
Homebrew/emacs: 1.947931
Caskroom/versions: 2.104662
Homebrew/fuse: 2.151206Homebrew/apache: 2.1793

Homebrew/games: 2.254464
Homebrew/devel-only: 2.29351
```

then it's stuck ...

**Comment by zmwangx** [comment]
No idea. I just tried this on a clean VM with nothing tapped other than core, and still got very reasonable results. Caskroom/cask took 3.560943s. You may attempt to step through. Since I can't reproduce I'm afraid there's nothing I can help with.

**Comment by cscetbon** [comment]
Should I try to uninstall/reinstall homebrew ?
Seeing all those times means it worked right ? I could I know where it is stuck except by stepping through ?  Isn't there a debug/super-verbose option ?

**Comment by zmwangx** [comment]
> Isn't there a debug/super-verbose option?

I don't think there is.

> Should I try to uninstall/reinstall homebrew?

Before that I would

```
cd "$(brew --repo)" && git fetch && git reset --hard origin/master && brew update
```

If that doesn't help and search important enough for you to go through the hassle.

It would be nice if you could try to install Homebrew in a clean VM though, just to make sure it's not some weird configuration that's getting in the way.

**Comment by cscetbon** [comment]
@zmwangx, what do you mean by : 
> It would be nice if you could try to install Homebrew in a clean VM though, just to make sure it's not some weird configuration that's getting in the way.

I'd like to do it, but how can I run a clean VM of OS X of my Mac without using virtual box and needing a lot of memory ? 
I don't think there is a docker equivalent to run OS X container on OS X ...

**Comment by zmwangx** [comment]
> but how can I run a clean VM of OS X of my Mac without using virtual box and needing a lot of memory?

You can use VMware Fusion too, but that's apparently not what you were asking. Short answer: you can't.

**Comment by zmwangx** [comment]
You can also try a clean installation of brew to a different prefix.

**Comment by cscetbon** [comment]
So I was able to reproduce it on a fresh install :
```
$ cd ~/tmp
$ git clone https://github.com/Homebrew/homebrew.git
Cloning into 'homebrew'...
remote: Counting objects: 346325, done.
remote: Compressing objects: 100% (2/2), done.
Receiving objects: 100% (346325/346325), 67.16 MiB | 1.00 MiB/s, done.
remote: Total 346325 (delta 1), reused 0 (delta 0), pack-reused 346323
Resolving deltas: 100% (254693/254693), done.
$ export PATH=$PWD/homebrew/bin:$PATH
$ brew search gnu
deja-gnu                 gnu-cobol                gnu-prolog               gnu-tar                  gnu-which                gnupg2                   gnutls
feedgnuplot              gnu-complexity           gnu-scientific-library   gnu-time                 gnumeric                 gnuplot                  gtk-gnutella
gnu-apl                  gnu-getopt               gnu-sed                  gnu-typist               gnupg                    gnuradio
gnu-barcode              gnu-indent               gnu-smalltalk            gnu-units                gnupg-pkcs11-scd         gnustep-make
homebrew/emacs/gnugo-emacs          homebrew/games/gnu-shogi            homebrew/php/php54-gnupg            homebrew/science/gnudatalanguage    Caskroom/cask/gnubg
homebrew/games/gnu-chess            homebrew/games/gnuski               homebrew/php/php55-gnupg            homebrew/versions/gnupg21           Caskroom/cask/gnucash
homebrew/games/gnu-go               homebrew/php/php53-gnupg            homebrew/php/php56-gnupg            homebrew/versions/gnuplot4
$ brew update
remote: Counting objects: 923, done.
remote: Compressing objects: 100% (803/803), done.
remote: Total 923 (delta 231), reused 530 (delta 91), pack-reused 0
Receiving objects: 100% (923/923), 852.89 KiB | 1.05 MiB/s, done.
Resolving deltas: 100% (231/231), completed with 159 local objects.
From ssh://github.com/Homebrew/brew
 + 73a2e75...a54c968 master     -> origin/master  (forced update)
 * [new tag]         0.1        -> 0.1
HEAD is now at a54c968 Merge pull request #977 from MikeMcQuaid/update-test-handle-missing-master
==> Homebrew has enabled anonymous aggregate user behaviour analytics
Read the analytics documentation (and how to opt-out) here:
  https://git.io/brew-analytics
Already up-to-date.
$ brew search gnu
deja-gnu                 gnu-cobol                gnu-prolog               gnu-tar                  gnu-which                gnupg2                   gnutls
feedgnuplot              gnu-complexity           gnu-scientific-library   gnu-time                 gnumeric                 gnuplot                  gtk-gnutella
gnu-apl                  gnu-getopt               gnu-sed                  gnu-typist               gnupg                    gnuradio
gnu-barcode              gnu-indent               gnu-smalltalk            gnu-units                gnupg-pkcs11-scd         gnustep-make
##### STUCK HERE ##### 
```

**Comment by cscetbon** [comment]
hey @zmwangx, any news on that ?

**Comment by zmwangx** [comment]
Unfortunately no, I've been too busy on other things, and since I can't reproduce, it's not very actionable for me and I still suspect it's a network problem somehow. Otherwise we probably have gotten more reports.

**Comment by cscetbon** [comment]
@zmwangx did you see how I've been able to reproduce it with another prefix ? Have you tried to do the same and you didn't get the issue ?

**Comment by zmwangx** [comment]
As a maintainer, not only do I have multiple Homebrew installations with different prefixes, I also have multiple VMs running multiple OS versions with Homebrew installed. I can reproduce nowhere, unfortunately.

**Comment by cscetbon** [comment]
:(

**Comment by mjhoy** [comment]
@cscetbon Anything you can glean from Activity Monitor or `lsof -i` while it's stuck?

**Comment by tonylotts** [comment]
I had this issue as well.  What fixed it for me, was to add the GitHub API token.  @cscetbon, perhaps some other app is making a lot of GitHub API calls, such that GitHub is throttling your account?

**Comment by tonylotts** [comment]
@cscetbon, I did see that you added the API token also.  My guess is that GitHub has multiple tiers to their throttling algorithm.

**Comment by cscetbon** [comment]
@tonylotts no it wasn't my issue. Since I upgraded to MacOS Sierra, the error is gone. I can't explain it though

**Comment by tonylotts** [comment]
For anyone's future reference, I had this issue on macOS Sierra.

**Comment by bayan9** [comment]
I"m having the exact same issue still

**Comment by ilovezfs** [comment]
Please open a new issue if you're still having problems.
